From: Chow Loong Jin <hyperair@debian.org>
Date: Mon, 9 Jan 2012 20:30:01 +0800
Subject: Fix mhash_deinit segfaults with snefru and whirlpool

mhash_deinit in lib/{whirlpool,snefru}.c would segfault when digest is
NULL, handle this case.  See http://bugs.debian.org/429180
---
 lib/snefru.c    | 3 +++
 lib/whirlpool.c | 4 ++++
 2 files changed, 7 insertions(+)

diff --git a/lib/snefru.c b/lib/snefru.c
index 6c10e22..e699c0e 100644
--- a/lib/snefru.c
+++ b/lib/snefru.c
@@ -859,6 +859,9 @@ static void snefru_digest(__const struct snefru_ctx *ctx, mutils_word8 *digest,
 {
 	mutils_word32 i;
 
+        if (!digest)
+            return;
+
 	for (i = 0; i < len; i++, digest += 4)
 	{
 		*(mutils_word32 *)digest = mutils_bend2sys32(ctx->hash[i]);
diff --git a/lib/whirlpool.c b/lib/whirlpool.c
index 36e36df..6eace51 100644
--- a/lib/whirlpool.c
+++ b/lib/whirlpool.c
@@ -970,6 +970,10 @@ void whirlpool_digest(__const struct whirlpool_ctx * ctx,
                     mutils_word8 * digest)
 {
    mutils_word32 i;
+
+   if (!digest)
+       return;
+
     /*
      * return the completed message digest:
      */
